### PR TITLE
Join.h refactoring: remove useless specialisations

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

**Comment by 4ertus2** [comment]
@amosbird, could you check all the changes are correct, please.

File: dbms/src/Interpreters/Join.h [hunk 1]
```diff
@@ -25,6 +25,45 @@
 
 namespace DB
 {
+
+namespace JoinStuff
+{
+/** Depending on template parameter, adds or doesn't add a flag, that element was used (row was joined).
+    * Depending on template parameter, decide whether to overwrite existing values when encountering the same key again
```
Comment (amosbird) [inline dbms/src/Interpreters/Join.h]: Comment needs update

**Comment by amosbird** [comment]
Yeah, it seems reasonable to use a normal variable instead of a template one. I was following the design pattern that "We don't pay what we don't use" as `join_overwrite` is a rarely used flag and somehow only makes sense when using it as a dictionary. But it does introduce code bloat. I guess some perf test might be needed to make sure that ordinary any joins won't suffer from that additional `join.anyTakeLastRow()` check?

**Comment by 4ertus2** [comment]
> I guess some perf test might be needed to make sure that ordinary any joins won't suffer from that additional `join.anyTakeLastRow()` check?

Compiler should be clever enough to place bool value instead this call. And CPU will predict that it's always the same. Of course, we need regression tests for joins performance. And some day we write them all :)

**Pull request merged**